sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,380 @@
+
+    self._header_lines_per_file = {}
+    self._file_pattern = file_pattern
+    self._compression_type = compression_type
+    self._buffer_size = buffer_size
+    self._validate = validate
+
+  def read_records(self, file_name, range_tracker):
+    def line_generator():
+      header_lines = []
+
+      def store_header_lines(header):
+        for line in header:
+          header_lines.append(line)
+
+      text_source = TextSource(
+          self._file_pattern,
+          0,  # min_bundle_size
+          self._compression_type,
+          strip_trailing_newlines=True,
+          coder=coders.StrUtf8Coder(),
+          buffer_size=self._buffer_size,
+          validate=self._validate,
+          skip_header_lines=0,
+          header_matcher_predicate=lambda x: x.startswith('#'),
+          header_processor=store_header_lines)
+
+      records = text_source.read_records(file_name, range_tracker)
+      header_processed = False
+      for line in records:
+        if not header_processed and header_lines:
+          for header in header_lines:
+            yield header
+          header_processed = True
+        # PyVCF has explicit str() calls when parsing INFO fields, which fails
+        # with UTF-8 decoded strings. Encode the line back to UTF-8.
+        yield line.encode('utf-8')
+
+    try:
+      vcf_reader = vcf.Reader(fsock=line_generator())
+      while True:
+        record = next(vcf_reader)
+        yield self._convert_to_variant_record(
+            record, vcf_reader.infos, vcf_reader.formats)
+    except SyntaxError as e:
+      raise ValueError('Invalid VCF header: %s' % str(e))
+    except (LookupError, ValueError) as e:
+      # TODO: Add 'strict' and 'loose' modes to not throw an
+      # exception in case of such failures.
+      raise ValueError('Invalid record in VCF file. Error: %s' % str(e))
+
+  def _convert_to_variant_record(self, record, infos, formats):
+    """Converts the PyVCF record to a :class:`Variant` object.
+
+    Args:
+      record (:class:`~vcf.model._Record`): An object containing info about a
+        variant.
+      infos (dict): The PyVCF dict storing INFO extracted from the VCF header.
+        The key is the info key and the value is :class:`~vcf.parser._Info`.
+      fromats (dict): The PyVCF dict storing FORMAT extracted from the VCF
+        header. The key is the FORMAT key and the value is
+        :class:`~vcf.parser._Format`.
+    Returns:
+      A :class:`Variant` object from the given record.
+    """
+    variant = Variant()
+    variant.reference_name = record.CHROM
+    variant.start = record.start
+    variant.end = record.end
+    variant.reference_bases = (
+        record.REF if record.REF != MISSING_FIELD_VALUE else None)
+    # ALT fields are classes in PyVCF (e.g. Substitution), so need convert them
+    # to their string representations.
+    variant.alternate_bases.extend(
+        [str(r) for r in record.ALT if r] if record.ALT else [])
```

Comment:
<!--new_thread; commit:70ade9d4fe6a8e31ff0be4d39b731510c93b4bf9; resolved:0-->
Please make sure that all lines here (for generating a Variant object) are covered by tests.